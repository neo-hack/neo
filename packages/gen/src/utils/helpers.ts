import { Workflow } from '../interface'
import { hooks } from '../utils/hooks'

import Observable from 'zen-observable'
import type Listr from 'listr'
import { LIFE_CYCLES } from 'src/constants'

/**
 * @description convert schema to listr
 */
export const toListr = (schema: Workflow) => {
  const tasks: Listr.ListrTask[] = Object.keys(schema.jobs || {}).map((key) => {
    const taskName = schema.jobs![key].name || key
    return {
      title: taskName,
      task: () => {
        let ob: ZenObservable.SubscriptionObserver<any>
        const OB = new Observable<any>((observer) => {
          ob = observer
        })
        hooks.addHooks({
          [LIFE_CYCLES.STEP]: (args: { step: string }) => {
            ob.next(args.step)
          },
          [taskName]: () => {
            ob.complete()
          },
        })
        return OB as any
      },
    }
  })
  return tasks
}
