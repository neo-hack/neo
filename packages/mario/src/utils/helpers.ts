import { Workflow } from '../interface'
import { hooks } from './hooks'

import Observable from 'zen-observable'
import type Listr from 'listr'
import { LIFE_CYCLES } from 'src/constants'

/**
 * @description convert schema to listr
 */
export const toListr = (schema: Workflow) => {
  const tasks: Listr.ListrTask[] = Object.keys(schema.jobs || {}).map((key) => {
    const taskName = schema.jobs![key].name || key
    return {
      title: taskName,
      task: () => {
        let ob: ZenObservable.SubscriptionObserver<any>
        const OB = new Observable<any>((observer) => {
          ob = observer
        })
        hooks.addHooks({
          [LIFE_CYCLES.STEP]: (args: { step: string }) => {
            ob.next(args.step)
          },
          [taskName]: (args: { error: Error }) => {
            if (!args.error) {
              ob.complete()
              return
            }
            ob.error(args.error)
          },
        })
        return OB as any
      },
    }
  })
  return tasks
}
