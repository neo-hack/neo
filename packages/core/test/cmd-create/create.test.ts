import path from 'path'
import fs from 'fs-extra'
import tempy from 'tempy'

import { execNeo } from '../helpers'

const storeDir = path.join(tempy.directory(), '.store')

describe('command create', () => {
  it('create on empty store', async () => {
    const destDir = tempy.directory()
    const { stdout } = await execNeo(['create', '--store-dir', storeDir], {
      cwd: destDir,
    })
    expect(stdout).toMatchSnapshot()
  }, 30000)

  it('create project from npm', async () => {
    const destDir = tempy.directory()
    await execNeo(['create', '@aiou/bin-template', 'target', '--store-dir', storeDir], {
      cwd: destDir,
    })
    expect(fs.existsSync(path.join(destDir, './target/README.md'))).toBe(true)
  }, 10000)

  it('create package from github should work', async () => {
    const destDir = tempy.directory()
    const url = 'https://github.com/neo-hack/bin-template'
    await execNeo(['create', url, 'target', '--store-dir', storeDir], {
      cwd: destDir,
    })
    expect(fs.existsSync(path.join(destDir, './target/README.md'))).toBe(true)
  }, 30000)
})

describe('command create postcreate', () => {
  it('postcreate clean up changelog file', async () => {
    const destDir = tempy.directory()
    await execNeo(['create', '@aiou/bin-template', 'target', '--store-dir', storeDir], {
      cwd: destDir,
    })
    expect(!fs.existsSync(path.join(destDir, './target/CHANGELOG.md'))).toBe(true)
  }, 10000)

  it('postcreate in monorepo clean up changelog, .eslintignore etc...', async () => {
    const destDir = tempy.directory()
    await execNeo(
      ['create', '@aiou/bin-template', 'target', '--store-dir', storeDir, '--mono', true],
      {
        cwd: destDir,
      },
    )
    const mono = ['.eslintignore', '.eslintrc', '.changeset', '.github', '.husky']
    const checkresult = mono.map((file) => !fs.existsSync(path.join(destDir, './target', file)))
    expect(checkresult.every((v) => v)).toBe(true)
  }, 10000)
})
