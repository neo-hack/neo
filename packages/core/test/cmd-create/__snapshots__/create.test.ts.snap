// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command create create on empty store 1`] = `
"NEO v3.5.0 (​https://github.com/neo-hack/neo​)

[log] No templates found, use follow commands load remote template or preset first
[log] 
Example:
$ neo add @aiou/preset-demo --preset
$ neo add @aiou/ts-lib-template
      "
`;
